src/App.tsx

```diff
@@ -1,16 +1,112 @@
-import React from 'react'
-import { Description } from './components/Description'
+import React, { useEffect, useRef, useState, useCallback } from 'react';
+import { Description } from './components/Description';
+
+import { Coords } from './types/types'
 
 function App () {
+  const containerRef = useRef<HTMLDivElement>(null);
+  const boxRef = useRef<HTMLDivElement>(null);
+
+  const [maxX, setMaxX] = useState(0);
+  const [maxY, setMaxY] = useState(0);
+  const [containerWidth, setContainerWidth] = useState(0);
+  const [containerHeight, setContainerHeight] = useState(0);
+
+  const isClicked = useRef(false);
+
+  const coords = useRef<Coords>({ startX: 0, startY: 0, lastX: 0, lastY: 0 });
+
+  const updateMaxPositions = useCallback(() => {
+    if (containerRef.current && boxRef.current) {
+      const { width: containerWidth, height: containerHeight } = containerRef.current.getBoundingClientRect();
+      const { width: boxWidth, height: boxHeight } = boxRef.current.getBoundingClientRect();
+
+      setMaxX(containerWidth - boxWidth + boxWidth / 2);
+      setMaxY(containerHeight - boxHeight + boxHeight / 2);
+      setContainerWidth(containerWidth);
+      setContainerHeight(containerHeight);
+
+      const clampedX = Math.min(Math.max(coords.current.lastX, (boxWidth / 2)), containerWidth - (boxWidth / 2));
+      const clampedY = Math.min(Math.max(coords.current.lastY, (boxHeight / 2)), containerHeight - (boxHeight / 2));
+
+      const boxStyle = boxRef.current.style;
+      boxStyle.left = `${clampedX}px`;
+      boxStyle.top = `${clampedY}px`;
+
+      coords.current = {
+        ...coords.current,
+        lastX: clampedX,
+        lastY: clampedY,
+      };
+    }
+  }, []);
+
+  const onMouseDown = useCallback((e: MouseEvent) => {
+    isClicked.current = true;
+    const { clientX, clientY } = e;
+    const { offsetLeft, offsetTop } = boxRef.current!;
```

Comment:
Suppressing type here is not best choice, also you can access to the element from the event.